### PR TITLE
Run rules for specific paths

### PR DESCRIPTION
**Pull request by arturbosch**
Introduces option to run rules only for specific paths which can be specified by the new 'exclusions' and 'inclusions' properties:
```
ShouldNotRunOnLibraryRule:
    active: true
    exclusions: "**/detekt/api/**"
    inclusions: "**/api/internal/**"
```

The Rule 'ShouldNotRunOnLibraryRule' (example rule) will only run on KtFile's from the `detekt-api` module inside the `internal` package.
This allows to implement for example #462 .

Implementation details:
- Basically wraps the FileSystem.getPathMatcher but only supports the globings part
- The regex feature is not os-independent. Globs conversion to regex is Windows aware

### CODE REVIEW

**Comment by schalkms** [comment]
@arturbosch can you please ping me when this pr is finished and working.
I'll happily review this pr then.

**Comment by arturbosch** [comment]
@schalkms thanks! the pr should work now. The support for exclusions and inclusions on the rule set level should be in a separate PR if needed, e.g. when replacing the test-pattern.

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -9,6 +9,7 @@
 - We use [Spek](https://github.com/spekframework/spek) for testing. Please use the `Spec.kt`-Suffix.
 - Feel free to add your name to the contributors list at the end of the readme file when opening a pull request.
 - The code in `detekt-api` and any rule in `detekt-rules` must be documented. We generate documentation for our website based on this modules.
+- If some Kotlin code in `resources` folder (like `detekt-formatting`) shows compilation error, right click on it and use `Mark as plain text`.
```
Comment (schalkms) [inline .github/CONTRIBUTING.md]: If some Kotlin code in the `resources` folder (like `detekt-formatting`) shows a compilation error, right click on it and use `Mark as plain text`.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Rule.kt [hunk 2]
```diff
@@ -45,7 +49,56 @@ abstract class Rule(
      */
     open val defaultRuleIdAliases: Set<String> = emptySet()
 
-    override fun visitCondition(root: KtFile): Boolean = active && !root.isSuppressedBy(ruleId, aliases)
+    /**
+     * When specified this rule only runs on [KtFile]'s with paths matching any inclusion pattern.
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Rule.kt]: Be aware of the [] syntax in comments. Please see #1572

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/SplitPattern.kt [hunk 3]
```diff
@@ -1,23 +1,29 @@
...
-            .map { it.removeSuffix("*") }
+        .splitToSequence(delimiters)
+        .map { it.trim() }
+        .filter { it.isNotBlank() }
+        .mapIf(removeTrailingAsterisks) { seq ->
+            seq.map { it.removePrefix("*") }
+                .map { it.removeSuffix("*") }
+        }.toList()
+
+    private fun <T> Sequence<T>.mapIf(
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/SplitPattern.kt]: Nice func

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/InclusionExclusionPatternsSpec.kt [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        }
+    }
+})
+
+private fun Path.runWith(rule: DummyRule): DummyRule {
+    rule.lint(this)
+    return rule
+}
+
+class OnlyLibraryTrackingRule(config: Config) : Rule(config) {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/InclusionExclusionPatternsSpec.kt]: internal?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/InclusionExclusionPatternsSpec.kt [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            counter++
+        }
+    }
+
+    fun assertOnlyLibraryFileVisited() {
+        assertThat(counter == 0 && libraryFileVisited).isTrue()
+    }
+}
+
+class DummyRule(config: Config = Config.empty) : Rule(config) {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/InclusionExclusionPatternsSpec.kt]: internal

File: detekt-rules/src/test/resources/cases/library/Dummy2.kt [hunk 6]
```diff
@@ -0,0 +1,2 @@
+package cases.library
+
```
Comment (schalkms) [inline detekt-rules/src/test/resources/cases/library/Dummy2.kt]: why is `Dummy2` the same as `Dummy`?

Comment (arturbosch) [reply detekt-rules/src/test/resources/cases/library/Dummy2.kt]: Just to see if in one test both dummies are visited and in other the library file. I extended the tests for this, Dummy2 was not used at all :).

**Comment by arturbosch** [comment]
Thx, I rebased and incoporated your feedback.
The documentation is want to write when the other related issues are finished to not have unpublished features documented and distract users (docu changes are immediately published ^^(

**Comment by schalkms** [comment]
@arturbosch I think you forgot to push the changes.

**Comment by arturbosch** [comment]
No not quite, just forgot to add the `a` in CONTRIBUTING.md :)

**Comment by schalkms** [comment]
Nice!
I'll do another review in the next three days, if I can squeeze it in.

**Comment by arturbosch** [comment]
Thanks man! I keep working on replacing `--filters` :)

**Pull request merged**